### ISSUE TITLE
[Tune] Pickling error with `WandbLoggerCallback` on Windows

### ISSUE THREAD

**Issue by ShaohanTian** [ray-project/ray#31366]
This is from the [official code](https://docs.ray.io/en/latest/tune/examples/tune-wandb.html), it works well on Linux systems but met errors on windows. All of the ray version is 2.2.0.
```python
import numpy as np

import ray
from ray import air, tune
from ray.air import session
from ray.air.integrations.wandb import setup_wandb
from ray.air.integrations.wandb import WandbLoggerCallback

def train_function(config):
    for i in range(30):
        loss = config["mean"] + config["sd"] * np.random.randn()
        session.report({"loss": loss})

def tune_with_callback():
    """Example for using a WandbLoggerCallback with the function API"""
    tuner = tune.Tuner(
        train_function,
        tune_config=tune.TuneConfig(
            metric="loss",
            mode="min"
        ),
        run_config=air.RunConfig(
            callbacks=[
                WandbLoggerCallback(project="Wandb_example")
            ]
        ),
        param_space={
            "mean": tune.grid_search([1, 2, 3, 4, 5]),
            "sd": tune.uniform(0.2, 0.8),
        },
    )
    tuner.fit()

if __name__ == '__main__':
    tune_with_callback()
```
And this if print log:
```txt
2022-12-30 11:50:41,732 INFO worker.py:1538 -- Started a local Ray instance.                                     
2022-12-30 11:50:46,508 INFO wandb.py:250 -- Already logged into W&B.                                            
Traceback (most recent call last):                                                                               
  File "C:\Users\shaohan.tian\scoop\apps\miniconda3\current\envs\steel\lib\site-packages\ray\tune\execution\trial_runner.py", line 928, in _wait_and_handle_event                                                                 
    self._on_pg_ready(next_trial)                                                                                
  File "C:\Users\shaohan.tian\scoop\apps\miniconda3\current\envs\steel\lib\site-packages\ray\tune\execution\trial_runner.py", line 1018, in _on_pg_ready                                                                          
    if not _start_trial(next_trial) and next_trial.status != Trial.ERROR:                                        
  File "C:\Users\shaohan.tian\scoop\apps\miniconda3\current\envs\steel\lib\site-packages\ray\tune\execution\trial_runner.py", line 1010, in _start_trial                                                                          
    self._callbacks.on_trial_start(                                                                              
  File "C:\Users\shaohan.tian\scoop\apps\miniconda3\current\envs\steel\lib\site-packages\ray\tune\callback.py", line 317, in on_trial_start                                                                                       
    callback.on_trial_start(**info)                                                                              
  File "C:\Users\shaohan.tian\scoop\apps\miniconda3\current\envs\steel\lib\site-packages\ray\tune\logger\logger.py", line 135, in on_trial_start                                                                                  
    self.log_trial_start(trial)                                                                                  
  File "C:\Users\shaohan.tian\scoop\apps\miniconda3\current\envs\steel\lib\site-packages\ray\air\integrations\wandb.py", line 527, in log_trial_start                                                                             
    self._trial_processes[trial].start()                                                                         
  File "C:\Users\shaohan.tian\scoop\apps\miniconda3\current\envs\steel\lib\multiprocessing\process.py", line 121, in start                                                                                                        
    self._popen = self._Popen(self)                                                                              
  File "C:\Users\shaohan.tian\scoop\apps\miniconda3\current\envs\steel\lib\multiprocessing\context.py", line 224, in _Popen                                                                                                       
    return _default_context.get_context().Process._Popen(process_obj)                                            
  File "C:\Users\shaohan.tian\scoop\apps\miniconda3\current\envs\steel\lib\multiprocessing\context.py", line 327, in _Popen                                                                                                       
    return Popen(process_obj)                                                                                    
  File "C:\Users\shaohan.tian\scoop\apps\miniconda3\current\envs\steel\lib\multiprocessing\popen_spawn_win32.py", line 93, in __init__                                                                                            
    reduction.dump(process_obj, to_child)                                                                        
  File "C:\Users\shaohan.tian\scoop\apps\miniconda3\current\envs\steel\lib\multiprocessing\reduction.py", line 60, in dump                                                                                                        
    ForkingPickler(file, protocol).dump(obj)                                                                     
  File "C:\Users\shaohan.tian\scoop\apps\miniconda3\current\envs\steel\lib\site-packages\ray\air\integrations\wandb.py", line 367, in __reduce__                                                                                  
    raise RuntimeError("_WandbLoggingProcess is not pickleable.")                                                
RuntimeError: _WandbLoggingProcess is not pickleable.                                                            
                                                                                                                 
During handling of the above exception, another exception occurred:                                              
                                                                                                                 
Traceback (most recent call last):                                                                               
  File "C:\Users\shaohan.tian\scoop\apps\miniconda3\current\envs\steel\lib\site-packages\ray\tune\tuner.py", line 272, in fit                                                                                                     
    return self._local_tuner.fit()                                                                               
  File "C:\Users\shaohan.tian\scoop\apps\miniconda3\current\envs\steel\lib\site-packages\ray\tune\impl\tuner_internal.py", line 420, in fit                                                                                       
    analysis = self._fit_internal(trainable, param_space)                                                        
  File "C:\Users\shaohan.tian\scoop\apps\miniconda3\current\envs\steel\lib\site-packages\ray\tune\impl\tuner_internal.py", line 532, in _fit_internal                                                                             
    analysis = run(                                                                                              
  File "C:\Users\shaohan.tian\scoop\apps\miniconda3\current\envs\steel\lib\site-packages\ray\tune\tune.py", line 
726, in run                                                                                                      
    runner.step()                                                                                                
  File "C:\Users\shaohan.tian\scoop\apps\miniconda3\current\envs\steel\lib\site-packages\ray\tune\execution\trial_runner.py", line 981, in step                                                                                   
    self._wait_and_handle_event(next_trial)                                                                      
  File "C:\Users\shaohan.tian\scoop\apps\miniconda3\current\envs\steel\lib\site-packages\ray\tune\execution\trial_runner.py", line 960, in _wait_and_handle_event                                                                 
    raise TuneError(traceback.format_exc())                                                                      
ray.tune.error.TuneError: Traceback (most recent call last):                                                     
  File "C:\Users\shaohan.tian\scoop\apps\miniconda3\current\envs\steel\lib\site-packages\ray\tune\execution\trial_runner.py", line 928, in _wait_and_handle_event                                                                 
    self._on_pg_ready(next_trial)                                                                                
  File "C:\Users\shaohan.tian\scoop\apps\miniconda3\current\envs\steel\lib\site-packages\ray\tune\execution\trial_runner.py", line 1018, in _on_pg_ready                                                                          
    if not _start_trial(next_trial) and next_trial.status != Trial.ERROR:                                        
  File "C:\Users\shaohan.tian\scoop\apps\miniconda3\current\envs\steel\lib\site-packages\ray\tune\execution\trial_runner.py", line 1010, in _start_trial                                                                          
    self._callbacks.on_trial_start(                                                                              
  File "C:\Users\shaohan.tian\scoop\apps\miniconda3\current\envs\steel\lib\site-packages\ray\tune\callback.py", line 317, in on_trial_start                                                                                       
    callback.on_trial_start(**info)                                                                              
  File "C:\Users\shaohan.tian\scoop\apps\miniconda3\current\envs\steel\lib\site-packages\ray\tune\logger\logger.py", line 135, in on_trial_start                                                                                  
    self.log_trial_start(trial)                                                                                  
  File "C:\Users\shaohan.tian\scoop\apps\miniconda3\current\envs\steel\lib\site-packages\ray\air\integrations\wandb.py", line 527, in log_trial_start                                                                             
    self._trial_processes[trial].start()                                                                         
  File "C:\Users\shaohan.tian\scoop\apps\miniconda3\current\envs\steel\lib\multiprocessing\process.py", line 121, in start                                                                                                        
    self._popen = self._Popen(self)                                                                              
  File "C:\Users\shaohan.tian\scoop\apps\miniconda3\current\envs\steel\lib\multiprocessing\context.py", line 224, in _Popen                                                                                                       
    return _default_context.get_context().Process._Popen(process_obj)                                            
  File "C:\Users\shaohan.tian\scoop\apps\miniconda3\current\envs\steel\lib\multiprocessing\context.py", line 327, in _Popen                                                                                                       
    return Popen(process_obj)                                                                                    
  File "C:\Users\shaohan.tian\scoop\apps\miniconda3\current\envs\steel\lib\multiprocessing\popen_spawn_win32.py", line 93, in __init__                                                                                            
    reduction.dump(process_obj, to_child)                                                                        
  File "C:\Users\shaohan.tian\scoop\apps\miniconda3\current\envs\steel\lib\multiprocessing\reduction.py", line 60, in dump                                                                                                        
    ForkingPickler(file, protocol).dump(obj)                                                                     
  File "C:\Users\shaohan.tian\scoop\apps\miniconda3\current\envs\steel\lib\site-packages\ray\air\integrations\wandb.py", line 367, in __reduce__                                                                                  
    raise RuntimeError("_WandbLoggingProcess is not pickleable.")                                                
RuntimeError: _WandbLoggingProcess is not pickleable.                                                            
                                                                                                                 
                                                                                                                 
The above exception was the direct cause of the following exception:                                             
                                                                                                                 
Traceback (most recent call last):                                                                               
  File "C:\Users\shaohan.tian\Desktop\laptop_python_run\wandb_test.py", line 36, in <module>                     
    tune_with_callback()                                                                                         
  File "C:\Users\shaohan.tian\Desktop\laptop_python_run\wandb_test.py", line 33, in tune_with_callback           
    tuner.fit()                                                                                                  
  File "C:\Users\shaohan.tian\scoop\apps\miniconda3\current\envs\steel\lib\site-packages\ray\tune\tuner.py", line 274, in fit                                                                                                     
    raise TuneError(                                                                                             
ray.tune.error.TuneError: The Ray Tune run failed. Please inspect the previous error messages for a cause. After 
fixing the issue, you can restart the run from scratch or continue this run. To continue this run, you can use `tuner = Tuner.restore("C:\Users\shaohan.tian\ray_results\train_function_2022-12-30_11-50-36")`.                   
Traceback (most recent call last):                                                                               
  File "<string>", line 1, in <module>                                                                           
  File "C:\Users\shaohan.tian\scoop\apps\miniconda3\current\envs\steel\lib\multiprocessing\spawn.py", line 116, in spawn_main                                                                                                     
    exitcode = _main(fd, parent_sentinel)                                                                        
  File "C:\Users\shaohan.tian\scoop\apps\miniconda3\current\envs\steel\lib\multiprocessing\spawn.py", line 126, in _main                                                                                                          
    self = reduction.pickle.load(from_parent)                                                                    
EOFError: Ran out of input                                                                                       
```
how to solve it?

**Comment by bveeramani** [comment]
Hey @ShaohanTian, thanks for reporting this issue.

What Python version are you using?

**Comment by bveeramani** [comment]
Possibly relevant: https://stackoverflow.com/questions/9670926/multiprocessing-on-windows-breaks

**Comment by ShaohanTian** [comment]
@bveeramani I'm using python==3.9, it doesn't solve my question with 
> Possibly relevant: https://stackoverflow.com/questions/9670926/multiprocessing-on-windows-breaks

**Comment by tjboller** [comment]
+1 but Im having the same issue with a mac using python 3.10.6

**Comment by ShaohanTian** [comment]
Have you solved it? @tjboller

**Comment by tjboller** [comment]
> Have you solved it? @tjboller 

No I just moved to the other options that were laid out in the docs

**Comment by bveeramani** [comment]
@tjboller @ShaohanTian does the nightly version of Ray work? We recently updated the `WandbLoggerCallback` implementation, and it should fix the issue (https://github.com/ray-project/ray/pull/30165)

**Comment by ShaohanTian** [comment]
Thanks for your reply @bveeramani; it works well when I use daily releases (Nightlies) on windows.
`pip install -U LINK_TO_WHEEL.whl`

**Comment by richardliaw** [comment]
Closing this now as it should be in 2.2.

**Comment by George-Dammous** [comment]
it is fixed in ray 2.3.0 version
